projects/admin/src/app/circulation/patron-transaction.service.ts

```diff
@@ -0,0 +1,248 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { BehaviorSubject, Observable} from 'rxjs';
+import { Injectable } from '@angular/core';
+import { PatronService } from '../service/patron.service';
+import {
+  PatronTransaction,
+  PatronTransactionEvent,
+  PatronTransactionEventType,
+  PatronTransactionStatus
+} from './patron-transaction';
+import { RecordService } from '@rero/ng-core';
+import { RouteToolService } from '../routes/route-tool.service';
+import { ToastrService } from 'ngx-toastr';
+import { TranslateService } from '@ngx-translate/core';
+import { UserService } from '../service/user.service';
+import { map } from 'rxjs/operators';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PatronTransactionService {
+
+  /** subject containing current loaded PatronTransactions */
+  patronTransactionsSubject$: BehaviorSubject<Array<PatronTransaction>> = new BehaviorSubject([]);
+
+  constructor(
+    private _recordService: RecordService,
+    private _userService: UserService,
+    private _routeToolService: RouteToolService,
+    private _patronService: PatronService,
+    private _toastService: ToastrService,
+    private _translateService: TranslateService
+  ) {}
+
+  /**
+   * Allow to build the query to send through the API to retrieve desired data
+   * @param patronPid: the patron pid
+   * @param loanPid: the loan pid
+   * @param type: the patron transaction main type
+   * @param status: the patron transaction status
+   * @return: a string representing the query to used based on function arguments
+   */
+  private _buildQuery(patronPid?: string, loanPid?: string, type?: string, status?: string): string {
+    const params: Array<string> = [];
+    if (patronPid !== undefined) {
+      params.push(`patron.pid:${patronPid}`);
+    }
+    if (loanPid !== undefined) {
+      params.push(`loan.pid:${loanPid}`);
+    }
+    if (type !== undefined) {
+      params.push(`type:${type}`);
+    }
+    if (status !== undefined) {
+      params.push(`status:${status}`);
+    }
+    return params.join(' AND ');
+  }
+
+  /** Load all patron-transactions corresponding to query parameter
+   *  @param query: string - The query used to retrieve patron-transactions
+   *  @param sort: string - The field used to sort the patron-transactions
+   */
+  private _loadPatronTransactions(query: string, sort: string = 'creation_date'): Observable<Array<PatronTransaction>> {
+    return this._recordService.getRecords(
+      'patron_transactions',
+      query,
+      1,
+      RecordService.MAX_REST_RESULTS_SIZE,
+      undefined,
+      undefined,
+      undefined,
```

Comment:
null instead of undefined, even if it changes nothing in this case.

Reply:
using 'null' cause problems with `ng-core/getRecords` : "Cannot used 'forEach' with null value"